collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java

```diff
@@ -265,269 +169,102 @@ private void deleteFileOrDir(String fileName) {
-                                        .getColumnIndex(FormsColumns.FORM_FILE_PATH)));
-                                deleteFileOrDir(formFilePath);
-                                deleteFileOrDir(storagePathProvider.getAbsoluteFormFilePath(del.getString(del
-                                        .getColumnIndex(FormsColumns.FORM_MEDIA_PATH))));
-                            } while (del.moveToNext());
-                        }
-                    } finally {
-                        if (del != null) {
-                            del.close();
-                        }
-                    }
-                    count = db.delete(FORMS_TABLE_NAME, where, whereArgs);
-                    break;
-
-                case FORM_ID:
-                    String formId = uri.getPathSegments().get(1);
-
-                    Cursor c = null;
-                    try {
-                        c = this.query(uri, null, where, whereArgs, null);
-                        // This should only ever return 1 record.
-                        if (c != null && c.getCount() > 0) {
-                            c.moveToFirst();
-                            do {
-                                deleteFileOrDir(storagePathProvider.getAbsoluteCacheFilePath(c.getString(c
-                                        .getColumnIndex(FormsColumns.JRCACHE_FILE_PATH))));
-                                String formFilePath = storagePathProvider.getAbsoluteFormFilePath(c.getString(c
-                                        .getColumnIndex(FormsColumns.FORM_FILE_PATH)));
-                                deleteFileOrDir(formFilePath);
-                                deleteFileOrDir(storagePathProvider.getAbsoluteFormFilePath(c.getString(c
-                                        .getColumnIndex(FormsColumns.FORM_MEDIA_PATH))));
-
-                                try {
-                                    // get rid of the old tables
-                                    ItemsetDbAdapter ida = new ItemsetDbAdapter();
-                                    ida.open();
-                                    ida.delete(c.getString(c
-                                            .getColumnIndex(FormsColumns.FORM_MEDIA_PATH))
-                                            + "/itemsets.csv");
-                                    ida.close();
-                                } catch (Exception e) {
-                                    // if something else is accessing the provider this may not exist
-                                    // so catch it and move on.
-                                }
-
-                            } while (c.moveToNext());
-                        }
-                    } finally {
-                        if (c != null) {
-                            c.close();
-                        }
+        deferDaggerInit();
+
+        int count;
+
+        FormDeleter formDeleter = new FormDeleter(formsRepository, instancesRepository, fastExternalItemsetsRepository);
```

Comment:
Shouldn't we inject this like other dependencies?

Reply:
This is a question I'm always asking myself and I only really have fuzzy answers. There's always a trade-off around what "level" of dependency is inverted (and then injected) especially when you're using a DI framework as there is some overhead in adding or changing a dependency (changing the constructor for instance). 

Generally I try and have my tests tell me what should be explicitly exposed as a dependency and what should be constructed from other dependencies. Here we test the `FormsProvider` behaviour as an "opaque box" - we just test its behaviour rather than it's interactions with other components, so I never end up writing a test that wants to check that `FormDeleter` is called (which would force me to make it an explicit dependency). This means there's no test that guarantees I'm using `FormDeleter`, although the part of `FormDeleter`'s behaviour we want is tested (deleting itemsets). 

All that is mostly out loud thoughts rather than an answer or a solid opinion, but I guess I would say that I wouldn't bother making it an explicit dependency until it would be easier to write a test if it was one. Does that make sense or is just rambling 😅?